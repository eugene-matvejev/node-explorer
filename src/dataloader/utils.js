/**
 * @param {Number[]|String[]} keys
 * @param {Number|String} key
 * @param {Boolean} aggregateInArray
 * @param {Object[]} values
 *
 * @returns {Object[]|Object[][]}
 *
 * examples:
 *  - processSQLResult(
 *      [1, 3, 2 ],
 *      'id',
 *      [
 *          { id: 1 },
 *          { id: 2 },
 *      ]
 *  )
 *  will return ->
 *  [
 *      { id: 1 },
 *      null,
 *      { id: 2 }
 *  ]
 *
 *  - processSQLResult(
 *      [1, 3, 2 ],
 *      'id',
 *      [
 *          { id: 1 },
 *          { id: 2 },
 *          { id: 2 },
 *      ]
 *  )
 *  will return ->
 *  [
 *      [
 *          { id: 1 },
 *      ],
 *      null,
 *      [
 *          { id: 2 },
 *          { id: 2 },
 *      ],
 *  ]
 *
 */
export const processSQLResult = (keys, key, values, aggregateInArray) => {
    const acc = {};

    for (const v of values) {
        if (!v[key]) {
            continue;
        }

        if (aggregateInArray) {
            !acc[v[key]]
                ? acc[v[key]] = [v]
                : acc[v[key]].push(v);
        } else acc[v[key]] = v;
    }

    return keys.map((v) => acc[v] || null);
}
