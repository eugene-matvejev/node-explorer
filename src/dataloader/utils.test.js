import { processSQLResult } from './utils';

describe('dataloader utils', () => {
    describe('processSQLResult', () => {
        it('should return Array<Object>, ordered by key value, and injecting NULL in absense of/falsy value', () => {
            const keys = [1, 2, 3, 4];
            const values = [
                { v: 1 },
                { v: 3 }, /** should be returned as 3rd */
                { v: 2 }, /** should be returned as 2nd */
                { v: null },
            ];

            expect(processSQLResult(keys, 'v', values)).toMatchSnapshot();
        });

        it('should return Array<Array<Object>>, ordered by key value, and injecting NULL in absense/falsy value', () => {
            const keys = [1, 2, 3];
            const values = [
                { v: 1 },
                { v: 3 }, /** should be returned in 3rd array */
                { v: 3 }, /** should be returned in 3rd array */
            ];

            expect(processSQLResult(keys, 'v', values, true)).toMatchSnapshot();
        });

        it('should return Array<Array<Object>>, ordered by key value, and injecting NULL in absense of a value [2nd and 3rd values should duplicated, because of key value]', () => {
            const keys = [1, 3, 3];
            const values = [
                { v: 1 },
                { v: 3 }, /** should be returned in 2nd and 3rd array */
                { v: 3 }, /** should be returned in 2nd and 3rd array */
            ];

            expect(processSQLResult(keys, 'v', values, true)).toMatchSnapshot();
        });
    });
});
