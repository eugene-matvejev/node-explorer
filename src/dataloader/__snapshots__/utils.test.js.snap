// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataloader utils processSQLResult should return Array<Array<Object>>, ordered by key value, and injecting NULL in absense of a value [2nd and 3rd values should duplicated, because of key value] 1`] = `
Array [
  Array [
    Object {
      "v": 1,
    },
  ],
  Array [
    Object {
      "v": 3,
    },
    Object {
      "v": 3,
    },
  ],
  Array [
    Object {
      "v": 3,
    },
    Object {
      "v": 3,
    },
  ],
]
`;

exports[`dataloader utils processSQLResult should return Array<Array<Object>>, ordered by key value, and injecting NULL in absense/falsy value 1`] = `
Array [
  Array [
    Object {
      "v": 1,
    },
  ],
  null,
  Array [
    Object {
      "v": 3,
    },
    Object {
      "v": 3,
    },
  ],
]
`;

exports[`dataloader utils processSQLResult should return Array<Object>, ordered by key value, and injecting NULL in absense of/falsy value 1`] = `
Array [
  Object {
    "v": 1,
  },
  Object {
    "v": 2,
  },
  Object {
    "v": 3,
  },
  null,
]
`;
